File: asap-precompute-go/snapshot_cache.go
package precompute

import (
	"fmt"
	"sync"
)

// SnapshotCache stores per-series sketch payloads for delta encoding
// (outbound) and inbound delta-apply against cached upstream
// snapshots. Mirrors today's per-processor `snapshots map[string][]byte`
// (outbound) and `IngestState.sketch_snapshots` (inbound) with explicit
// per-direction methods.
//
// Two maps avoid cross-direction collisions: the same series key may
// be both produced (outbound) and consumed (inbound) by the same
// host when running as a forwarder, and the two byte streams are not
// interchangeable (a remote sender's snapshot is not what we'd emit
// locally).
type SnapshotCache struct {
	mu         sync.RWMutex
	outbound   map[string][]byte
	inbound    map[string][]byte
	generation uint64
	seen       map[string]uint64
}

// NewSnapshotCache constructs an empty cache.
func NewSnapshotCache() *SnapshotCache {
	return &SnapshotCache{
		outbound:   make(map[string][]byte),
		inbound:    make(map[string][]byte),
		generation: 1,
		seen:       make(map[string]uint64),
	}
}

// CacheOutbound stores the latest full sketch payload, keyed by
// seriesKey. Returns true if this is the first snapshot for that
// key (caller can use this to force a PROTO_FULL on next emit).
//
// Stores a defensive copy of payload so a caller mutating its slice
// after caching does not race the cache reader.
func (c *SnapshotCache) CacheOutbound(seriesKey string, payload []byte) (firstTime bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	_, existed := c.outbound[seriesKey]
	cp := make([]byte, len(payload))
	copy(cp, payload)
	c.outbound[seriesKey] = cp
	c.seen[seriesKey] = c.generation
	return !existed
}

// GetOutbound returns the cached outbound payload, or nil. The
// returned slice MUST NOT be mutated by the caller — it aliases the
// cache's storage. (Callers that need to mutate should copy first.)
func (c *SnapshotCache) GetOutbound(seriesKey string) []byte {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.outbound[seriesKey]
}

// CacheInbound stores an upstream snapshot for delta apply.
func (c *SnapshotCache) CacheInbound(seriesKey string, payload []byte) {
	c.mu.Lock()
	defer c.mu.Unlock()
	cp := make([]byte, len(payload))
	copy(cp, payload)
	c.inbound[seriesKey] = cp
	c.seen[seriesKey] = c.generation
}

// BeginGeneration advances the reusable liveness epoch used by rotation.
func (c *SnapshotCache) BeginGeneration() uint64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.generation++
	return c.generation
}

// Touch marks a retained series without allocating a per-rotation keep set.
func (c *SnapshotCache) Touch(seriesKey string, generation uint64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if generation == c.generation {
		c.seen[seriesKey] = generation
	}
}

// EndGeneration removes snapshots not touched by the closed generation. Cache
// writes racing with serialization mark themselves in the current generation.
func (c *SnapshotCache) EndGeneration(generation uint64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if generation != c.generation {
		return
	}
	for key := range c.outbound {
		if c.seen[key] != generation {
			delete(c.outbound, key)
		}
	}
	for key := range c.inbound {
		if c.seen[key] != generation {
			delete(c.inbound, key)
		}
	}
	for key, seen := range c.seen {
		if seen != generation {
			delete(c.seen, key)
		}
	}
}

// GetInbound returns the cached upstream snapshot or nil.
func (c *SnapshotCache) GetInbound(seriesKey string) []byte {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.inbound[seriesKey]
}

// emptyBaseDeltaSketch is implemented by sketches that opt in to
// "true per-window deltas" (delta-baseline-contract.md §3).
//
// DeltaAgainstEmptyBase returns the snapshot bytes the SnapshotCache
// caches as the outbound base AFTER each window-close emit — i.e. the
// snapshot of an EMPTY sketch of the same shape. The next window's
// ComputeDeltaAgainst then diffs against empty, so its delta is that
// window's own full per-window state encoded as a delta (no
// cross-window subtraction). This is what makes the backend's future
// per-window base rotation correct.
//
// DDSketch, CMS, CountSketch, and HLL implement this; KLL does NOT
// (it stays full-only), so KLL keeps the legacy always-refresh behavior
// via the type-assertion miss below. A family that opts in only
// conditionally returns a nil/empty base from DeltaAgainstEmptyBase to
// fall back to legacy refresh for that call (e.g. CMS msgpack mode).
type emptyBaseDeltaSketch interface {
	DeltaAgainstEmptyBase() ([]byte, error)
}

// ComputeDelta diffs current sketch state against the cached outbound
// snapshot for that seriesKey. Returns (payload, isFull, err) where
// isFull means the runtime should emit PROTO_FULL (either no prior
// snapshot existed, or the delta exceeded threshold).
//
// Base refresh — two modes:
//
//   - Legacy always-refresh (KLL, and any family that does not opt in):
//     every call updates the cached previous snapshot to the current
//     sketch state, so successive sub-threshold deltas are each computed
//     against the immediately preceding window. This matches the
//     established behavior of the legacy OTel sketch processors.
//   - Per-window deltas (DDSketch / CMS / CountSketch / HLL, via
//     emptyBaseDeltaSketch): after each window-close emit the cached base
//     is reset to the EMPTY-sketch snapshot, so the next window diffs
//     against empty and transmits its OWN per-window state as a delta (no
//     cross-window subtraction). See delta-baseline-contract.md §3.
//
// The Sketch interface's ComputeDeltaAgainst does the actual diff
// using the algorithm-specific delta-encoding rules from sketchlib-go.
func (c *SnapshotCache) ComputeDelta(
	seriesKey string,
	current Sketch,
	threshold uint64,
) (payload []byte, isFull bool, err error) {
	if current == nil {
		return nil, false, fmt.Errorf("compute delta: nil sketch")
	}
	c.mu.RLock()
	prev := c.outbound[seriesKey]
	c.mu.RUnlock()
	// Compute the wire payload first (full on first call or above
	// threshold; sparse delta otherwise).
	if prev == nil {
		// First time — emit full.
		full, snapErr := current.Snapshot()
		if snapErr != nil {
			return nil, false, fmt.Errorf("snapshot: %w", snapErr)
		}
		payload = full
		isFull = true
	} else {
		delta, full, dErr := current.ComputeDeltaAgainst(prev, threshold)
		if dErr != nil {
			return nil, false, fmt.Errorf("compute delta: %w", dErr)
		}
		payload = delta
		isFull = full
	}
	// Refresh the cached outbound base for the NEXT window's delta.
	//
	// Per-window deltas: if the sketch opts in via
	// emptyBaseDeltaSketch (DDSketch / CMS / CountSketch / HLL), reset
	// the cached base to the EMPTY-sketch snapshot after this
	// window-close emit, so the next window diffs against empty and
	// transmits its own per-window state as a delta — no cross-window
	// subtraction.
	//
	// A family may opt in only conditionally (e.g. CMS in proto mode but
	// not msgpack mode, which cannot carry deltas). DeltaAgainstEmptyBase
	// then returns a nil/empty base to signal "not this call": we fall
	// through to the legacy always-refresh below for that case.
	if eb, ok := current.(emptyBaseDeltaSketch); ok {
		emptyBase, ebErr := eb.DeltaAgainstEmptyBase()
		if ebErr != nil {
			return nil, false, fmt.Errorf("empty base: %w", ebErr)
		}
		if len(emptyBase) > 0 {
			c.CacheOutbound(seriesKey, emptyBase)
			return payload, isFull, nil
		}
	}
	// Legacy always-refresh: update the cached outbound to the latest
	// full snapshot so the next ComputeDelta call diffs against the
	// just-emitted window. When isFull=true the wire payload IS the
	// snapshot, so reuse it; otherwise serialize a fresh full snapshot
	// for the cache. Both branches end with c.outbound[seriesKey] ==
	// latest full state.
	if isFull {
		c.CacheOutbound(seriesKey, payload)
	} else {
		full, snapErr := current.Snapshot()
		if snapErr != nil {
			return nil, false, fmt.Errorf("snapshot: %w", snapErr)
		}
		c.CacheOutbound(seriesKey, full)
	}
	return payload, isFull, nil
}

// RetainKeys prunes the cache down to ONLY the keys present in retain,
// deleting every outbound/inbound entry whose key is not in the set.
//
// Without this the outbound/inbound maps only ever grow: with delta
// transmission every series key ever observed retains a snapshot copy
// forever, even after the series vanishes, pinning agent memory for the
// process lifetime (MaxSeries bounds only the live window map, not the
// cache). finishRotate calls this after each window-close emit with the
// just-closed window's key set, so a series that did not reappear this
// window has its cached snapshots evicted.
//
// A nil/empty retain set is treated as "keep nothing" and clears both
// maps — callers that mean "no series this window" want the cache empty.
func (c *SnapshotCache) RetainKeys(retain map[string]struct{}) {
	c.mu.Lock()
	defer c.mu.Unlock()
	for k := range c.outbound {
		if _, keep := retain[k]; !keep {
			delete(c.outbound, k)
		}
	}
	for k := range c.inbound {
		if _, keep := retain[k]; !keep {
			delete(c.inbound, k)
		}
	}
}

// Delete removes a single series key's cached outbound and inbound
// snapshots. Used when a series is explicitly evicted; a no-op when the
// key is absent.
func (c *SnapshotCache) Delete(seriesKey string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.outbound, seriesKey)
	delete(c.inbound, seriesKey)
	delete(c.seen, seriesKey)
}

// Reset clears all cached state (used in tests and on shutdown).
func (c *SnapshotCache) Reset() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.outbound = make(map[string][]byte)
	c.inbound = make(map[string][]byte)
	c.seen = make(map[string]uint64)
	c.generation++
}

// LenOutbound returns the number of cached outbound snapshots; for
// tests and the telemetry layer.
func (c *SnapshotCache) LenOutbound() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return len(c.outbound)
}

// LenInbound returns the number of cached inbound snapshots; for
// tests and the telemetry layer.
func (c *SnapshotCache) LenInbound() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return len(c.inbound)
}

// ComputeSubWindowDelta diffs the current (still-accumulating) sketch against
// the immediately-preceding emit for seriesKey and refreshes the cached base to
// the just-emitted full state — the always-refresh path applied WITHIN a window
// between sub-window emits. Unlike the boundary ComputeDelta it NEVER applies
// the per-window empty-base reset (that reset is the boundary action; doing it
// mid-window would re-ship the whole window-to-date and defeat the savings) and
// does NOT rotate/reset the sketch. The first emit of a window (no prior base)
// is full; thereafter sub-window emits are sparse deltas against the previous
// emit.
func (c *SnapshotCache) ComputeSubWindowDelta(
	seriesKey string,
	current Sketch,
	threshold uint64,
) (payload []byte, isFull bool, err error) {
	if current == nil {
		return nil, false, fmt.Errorf("compute sub-window delta: nil sketch")
	}
	c.mu.RLock()
	prev := c.outbound[seriesKey]
	c.mu.RUnlock()
	if prev == nil {
		full, snapErr := current.Snapshot()
		if snapErr != nil {
			return nil, false, fmt.Errorf("snapshot: %w", snapErr)
		}
		c.CacheOutbound(seriesKey, full)
		return full, true, nil
	}
	delta, full, dErr := current.ComputeDeltaAgainst(prev, threshold)
	if dErr != nil {
		return nil, false, fmt.Errorf("compute delta: %w", dErr)
	}
	// Always-refresh: cache the current full state as the base for the NEXT
	// sub-window emit so it diffs against THIS emit (incremental), not empty.
	if full {
		c.CacheOutbound(seriesKey, delta)
	} else {
		fullSnap, snapErr := current.Snapshot()
		if snapErr != nil {
			return nil, false, fmt.Errorf("snapshot: %w", snapErr)
		}
		c.CacheOutbound(seriesKey, fullSnap)
	}
	return delta, full, nil
}

File: asap-precompute-rs/src/snapshot_cache.rs
//! Per-series sketch payload cache for delta encoding (outbound) and
//! inbound delta-apply against cached upstream snapshots.
//!
//! Mirrors `asap-precompute-go/snapshot_cache.go` and today's
//! per-processor `snapshots map[string][]byte` (outbound) +
//! `IngestState::sketch_snapshots` (inbound).

use std::collections::HashMap;
use std::sync::RwLock;

use crate::precompute::{DeltaResult, PrecomputeError, Sketch};

/// Stores per-series sketch payloads for delta encoding (outbound)
/// and inbound delta-apply against cached upstream snapshots.
///
/// Mirrors Go `precompute.SnapshotCache`.
///
/// Two maps avoid cross-direction collisions: the same series key
/// may be both produced (outbound) and consumed (inbound) by the
/// same host when running as a forwarder, and the two byte streams
/// are not interchangeable (a remote sender's snapshot is not what
/// we'd emit locally).
///
/// **Always-refresh policy** (mirrors the Go side / ADR-0002
/// §"Behavior preservation"): every call to
/// [`Self::compute_delta`] updates the cached previous snapshot to
/// the current sketch state. Successive sub-threshold deltas are
/// therefore each computed against the immediately preceding
/// window, matching the established behavior of all five legacy
/// OTel sketch processors (DDSketch / KLL / HLL / CountSketch /
/// CountMinSketch). There is no configurable
/// "refresh-only-on-full" mode — that earlier design was a bug
/// because it forced downstream consumers to merge a chain of
/// deltas back to the original baseline rather than apply each
/// delta to the previous window's reconstructed state.
pub struct SnapshotCache {
    inner: RwLock<SnapshotCacheInner>,
}

struct SnapshotCacheInner {
    outbound: HashMap<String, Vec<u8>>,
    inbound: HashMap<String, Vec<u8>>,
}

impl Default for SnapshotCache {
    fn default() -> Self {
        Self::new()
    }
}

impl SnapshotCache {
    /// Constructs an empty cache. Mirrors Go `NewSnapshotCache`.
    pub fn new() -> Self {
        Self {
            inner: RwLock::new(SnapshotCacheInner {
                outbound: HashMap::new(),
                inbound: HashMap::new(),
            }),
        }
    }

    /// Stores the latest full sketch payload, keyed by `series_key`.
    /// Returns `true` if this is the first snapshot for that key
    /// (caller can use this to force a [`crate::envelope::Encoding::ProtoFull`]
    /// on next emit).
    ///
    /// Stores a defensive copy of `payload` so a caller mutating its
    /// slice after caching does not race the cache reader. Mirrors
    /// Go `(*SnapshotCache).CacheOutbound`.
    pub fn cache_outbound(&self, series_key: &str, payload: &[u8]) -> bool {
        let mut g = self.inner.write().expect("snapshot cache poisoned");
        let first_time = !g.outbound.contains_key(series_key);
        g.outbound.insert(series_key.to_string(), payload.to_vec());
        first_time
    }

    /// Returns the cached outbound payload, or `None`. Mirrors Go
    /// `(*SnapshotCache).GetOutbound`.
    pub fn get_outbound(&self, series_key: &str) -> Option<Vec<u8>> {
        self.inner
            .read()
            .expect("snapshot cache poisoned")
            .outbound
            .get(series_key)
            .cloned()
    }

    /// Stores an upstream snapshot for delta apply. Mirrors Go
    /// `(*SnapshotCache).CacheInbound`.
    pub fn cache_inbound(&self, series_key: &str, payload: &[u8]) {
        let mut g = self.inner.write().expect("snapshot cache poisoned");
        g.inbound.insert(series_key.to_string(), payload.to_vec());
    }

    /// Returns the cached upstream snapshot or `None`. Mirrors Go
    /// `(*SnapshotCache).GetInbound`.
    pub fn get_inbound(&self, series_key: &str) -> Option<Vec<u8>> {
        self.inner
            .read()
            .expect("snapshot cache poisoned")
            .inbound
            .get(series_key)
            .cloned()
    }

    /// Diffs current sketch state against the cached outbound
    /// snapshot for `series_key`.
    ///
    /// Returns a [`DeltaResult`] where `is_full` means the runtime
    /// should emit `ProtoFull` (either no prior snapshot existed,
    /// or the delta exceeded `threshold`). Mirrors Go
    /// `(*SnapshotCache).ComputeDelta`.
    ///
    /// Always-refresh: every call updates the cached previous
    /// snapshot to the current sketch state. When `is_full=true`
    /// the wire payload IS the full snapshot, so it is reused for
    /// the cache; otherwise a fresh full snapshot is serialized
    /// for the cache. Both branches end with the cache holding
    /// the latest full state.
    pub fn compute_delta(
        &self,
        series_key: &str,
        current: &dyn Sketch,
        threshold: u64,
    ) -> Result<DeltaResult, PrecomputeError> {
        let prev = self
            .inner
            .read()
            .expect("snapshot cache poisoned")
            .outbound
            .get(series_key)
            .cloned();

        let result = match prev {
            None => {
                // First time — emit full.
                let full = current.snapshot()?;
                DeltaResult {
                    payload: full,
                    is_full: true,
                }
            }
            Some(prev_bytes) => current.compute_delta_against(&prev_bytes, threshold)?,
        };

        // Refresh the cached outbound base for the NEXT window's delta.
        //
        // (delta-baseline-contract.md §3) — per-window deltas:
        // families that opt in via `Sketch::delta_against_empty_base`
        // (DDSketch this phase) reset the cached base to the EMPTY-sketch
        // snapshot after each window-close emit, so the next window diffs
        // against empty and transmits its OWN per-window state as a delta
        // (no cross-window subtraction). This is what makes the backend's
        // future per-window base rotation correct.
        //
        // Legacy always-refresh (all other families, `None` default):
        // update the cached base to the just-emitted full snapshot, so the
        // next delta is computed against the immediately preceding window.
        // When is_full=true the wire payload IS the snapshot; reuse it.
        // Otherwise serialize a fresh snapshot.
        match current.delta_against_empty_base()? {
            Some(empty_base) => {
                self.cache_outbound(series_key, &empty_base);
            }
            None => {
                if result.is_full {
                    self.cache_outbound(series_key, &result.payload);
                } else {
                    let full = current.snapshot()?;
                    self.cache_outbound(series_key, &full);
                }
            }
        }
        Ok(result)
    }

    /// Clears all cached state. Used in tests and on shutdown.
    /// Mirrors Go `(*SnapshotCache).Reset`.
    pub fn reset(&self) {
        let mut g = self.inner.write().expect("snapshot cache poisoned");
        g.outbound.clear();
        g.inbound.clear();
    }

    /// Returns the number of cached outbound snapshots. Mirrors Go
    /// `(*SnapshotCache).LenOutbound`.
    pub fn len_outbound(&self) -> usize {
        self.inner
            .read()
            .expect("snapshot cache poisoned")
            .outbound
            .len()
    }

    /// Returns the number of cached inbound snapshots. Mirrors Go
    /// `(*SnapshotCache).LenInbound`.
    pub fn len_inbound(&self) -> usize {
        self.inner
            .read()
            .expect("snapshot cache poisoned")
            .inbound
            .len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn outbound_first_time_true_then_false() {
        let c = SnapshotCache::new();
        assert!(c.cache_outbound("series-1", &[1, 2, 3]));
        assert!(!c.cache_outbound("series-1", &[4, 5, 6]));
        assert_eq!(c.len_outbound(), 1);
        assert_eq!(c.get_outbound("series-1"), Some(vec![4, 5, 6]));
    }

    #[test]
    fn inbound_independent_from_outbound() {
        let c = SnapshotCache::new();
        c.cache_outbound("k", b"out");
        c.cache_inbound("k", b"in");
        assert_eq!(c.get_outbound("k"), Some(b"out".to_vec()));
        assert_eq!(c.get_inbound("k"), Some(b"in".to_vec()));
        assert_eq!(c.len_outbound(), 1);
        assert_eq!(c.len_inbound(), 1);
    }

    #[test]
    fn reset_clears_all() {
        let c = SnapshotCache::new();
        c.cache_outbound("k", b"v");
        c.cache_inbound("k", b"v");
        c.reset();
        assert_eq!(c.len_outbound(), 0);
        assert_eq!(c.len_inbound(), 0);
    }

    #[test]
    fn cache_stores_defensive_copy() {
        let c = SnapshotCache::new();
        let mut payload = vec![1, 2, 3];
        c.cache_outbound("k", &payload);
        payload[0] = 99;
        // Cache must not have mutated.
        assert_eq!(c.get_outbound("k"), Some(vec![1, 2, 3]));
    }

    // ---- per-window delta tests (DDSketch) ----

    use crate::precompute::QuantileSketch;
    use crate::sketches::ddsketch::DDSketchWrapper;

    /// Build a fresh per-window DDSketch wrapper from a value range. The
    /// runtime resets per-series state every window, so each window is an
    /// independent sketch — mirrored here by a fresh wrapper per window.
    fn window_sketch(alpha: f64, values: impl IntoIterator<Item = f64>) -> DDSketchWrapper {
        let mut w = DDSketchWrapper::new(alpha);
        for v in values {
            w.update(v);
        }
        w
    }

    /// With delta mode on, applying the emitted DDSketch delta
    /// to an EMPTY base reconstructs the window's full state, and
    /// quantiles match within the α relative-accuracy bound.
    #[test]
    fn ddsketch_delta_against_empty_round_trips() {
        let alpha = 0.01;
        let c = SnapshotCache::new();

        // Window 1 — first emit is FULL (no prior base).
        let w1 = window_sketch(alpha, (1..=200).map(|i| i as f64));
        let r1 = c.compute_delta("series", &w1, 1).unwrap();
        assert!(r1.is_full, "first window must emit a full frame");

        // Window 2 — a fresh per-window sketch; emit must be a DELTA now
        // (the cache reset the base to empty at window-1 close).
        let w2 = window_sketch(alpha, (1..=200).map(|i| i as f64));
        let r2 = c.compute_delta("series", &w2, 1).unwrap();
        assert!(!r2.is_full, "window 2 must emit a delta, not a full frame");
        assert!(!r2.payload.is_empty(), "delta payload must be non-empty");

        // Apply the delta to an EMPTY base → reconstructs window 2.
        let mut recon = DDSketchWrapper::new(alpha);
        recon.apply_delta(&r2.payload).unwrap();
        assert_eq!(recon.inner().total_count(), w2.inner().total_count());

        // Quantiles match within α.
        for q in [0.5, 0.9, 0.99] {
            let got = recon.quantile(q);
            let want = w2.quantile(q);
            assert!(
                (got / want - 1.0).abs() <= alpha,
                "q={q}: recon={got} want={want}"
            );
        }
    }

    /// Two consecutive windows each emit their OWN state —
    /// window 2's delta is NOT diffed against window 1 (no cross-window
    /// subtraction). Even when window 1's events overlap window 2's, the
    /// emitted delta reconstructs window 2 exactly from empty.
    #[test]
    fn ddsketch_consecutive_windows_no_cross_window_subtraction() {
        let alpha = 0.01;
        let c = SnapshotCache::new();

        // Window 1: each value inserted TWICE (high counts).
        let mut w1 = DDSketchWrapper::new(alpha);
        for i in 1..=100 {
            w1.update(i as f64);
            w1.update(i as f64);
        }
        let r1 = c.compute_delta("s", &w1, 1).unwrap();
        assert!(r1.is_full);

        // Window 2: SAME value range but each inserted ONCE (lower counts).
        // If the cache diffed window 2 against window 1 (cross-window),
        // every bucket delta would saturate to 0 and the emitted delta
        // would be empty — under-transmitting window 2 entirely.
        let mut w2 = DDSketchWrapper::new(alpha);
        for i in 1..=100 {
            w2.update(i as f64);
        }
        let r2 = c.compute_delta("s", &w2, 1).unwrap();
        assert!(!r2.is_full);

        // Reconstruct window 2 from EMPTY — must equal window 2's own
        // full state, proving no cross-window subtraction occurred.
        let mut recon = DDSketchWrapper::new(alpha);
        recon.apply_delta(&r2.payload).unwrap();
        assert_eq!(
            recon.inner().total_count(),
            w2.inner().total_count(),
            "window 2 delta must carry window 2's full count"
        );

        // The reconstructed-from-empty state must match window 2's own
        // full per-bucket counts (compared on absolute bucket index —
        // the reconstructed store's backing-array padding differs from
        // the chunk-grown original, but the logical distribution is
        // identical: delta-against-empty == window 2's own state).
        let w2_sk = w2.inner();
        let r_sk = recon.inner();
        for (i, &c) in w2_sk.store_counts.iter().enumerate() {
            if c == 0 {
                continue;
            }
            let k = w2_sk.store_offset + i as i32;
            let r_idx = (k - r_sk.store_offset) as usize;
            assert_eq!(r_sk.store_counts[r_idx], c, "bucket k={k}");
        }

        // And it must NOT equal window 1's state (which had double counts).
        assert_ne!(recon.inner().total_count(), w1.inner().total_count());
    }
}
